docs/DeveloperGuide.adoc

```diff
@@ -767,16 +807,139 @@ Use case ends.
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 2a. The list is empty.
++
+Use case ends.
+
+* 2a. No such person was found.
++
+[none]
+** 2a1. AddressBook informs the user that no matching users were found.
++
+Use case ends.
+
+[discrete]
+=== Use case: Edit person
+
+*MSS*
+
+1.  User requests to list persons
+2.  AddressBook shows a list of persons
+3.  User requests to change the details of a specific person in the list
+4.  AddressBook changes the details of the specified person.
+5.  AddressBook shows the new details of the person.
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 2a. The list is empty.
++
+Use case ends.
+
+* 3a. The given index is invalid.
++
+[none]
+** 3a1. AddressBook shows an error message.
++
+Use case resumes at step 2.
+
+* 3a. The specified detail to change is invalid.
++
+[none]
+** 3a1. AddressBook shows an error message.
++
+Use case resumes at step 2.
+
+* 4a. The specified detail to change is exactly the same as the previous detail.
++
+[none]
+** 4a1. AddressBook informs the user that no change was made.
++
+Use case ends.
+
+[discrete]
+=== Use case: Undo/Redo
+
+*MSS*
+
+1.  User requests to undo/redo a specified number of steps
+2.  AddressBook undos/redos the last X commands, where X was the number of commands to undo/redo
+3.  AddressBook displays a success message
+4.  AddressBook displays a list of all the commands that were undone/redone.
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 2a. The number of commands entered were less than the specified number of undo commands.
++
+[none]
+** 2a1. AddressBook undos all previous commands.
++
+Use case resumes at step 3.
+
+[none]
+* 2a. The number of commands entered were less than the specified number of redo commands.
++
+[none]
+** 2a1. AddressBook redos all previously undone commands.
++
+Use case resumes at step 3
+
+[discrete]
+=== Use case: Sort
+
+*MSS*
+
+1.  User requests to sort by a criteria
+2.  AddressBook sorts the contacts by the criteria
+3.  AddressBook shows a success message.
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 2a. The list is empty.
++
+Use case ends.
+
+* 1a. The given criteria is invalid.
++
+[none]
+** 1a1. AddressBook shows an error message.
++
+Use case resumes at step 1.
 
 [appendix]
 == Non Functional Requirements
 
 .  Should work on any link:#mainstream-os[mainstream OS] as long as it has Java `1.8.0_60` or higher installed.
 .  Should be able to hold up to 1000 persons without a noticeable sluggishness in performance for typical usage.
 .  A user with above average typing speed for regular English text (i.e. not code, not system admin commands) should be able to accomplish most of the tasks faster using commands than using the mouse.
-
-{More to be added}
+.  Feature sets are to be rolled out on a regular weekly basis, following the proper forking workflow procedure.
+.  The application should be intuitive for all users, regardless of technical ability.
```

Comment:
This is too ambitious to be practical. It is safer to specify a lower bound. e.g. for those who use computers regularly. In any case, the product is targeted for those who prefer typing over mouse, so they are unlikely to be IT noobs.